### PR TITLE
Fix/adf 12/code integration

### PR DESCRIPTION
**Pull request by boajer**
https://oat-sa.atlassian.net/browse/ADF-12

### CODE REVIEW

File: config/index.conf.php [hunk 1]
```diff
@@ -0,0 +1,328 @@
...
+            ],
+            'settings' => [
+                'index' => [
+                    'number_of_shards' => '1',
+                    'number_of_replicas' => '1',
+                ],
+            ],
+        ],
+    ],
+];
```
Comment (bartlomiejmarszal) [inline config/index.conf.php:328]: /n

**Review by mike-ionut-mihai-sandu-tao — APPROVED** [review]
- [x] New code is covered by tests (if applicable)
- [ ] Tests are running successfully (old and new ones) on my local machine (if applicable)
- [x] New code is respecting code style rules
- [x] New code is respecting best practices
- [ ] New code is not subject to concurrency issues (if applicable)
- [ ] Feature is working correctly on my local machine (if applicable)
- [ ] Acceptance criteria are respected
- [ ] Pull request title and description are meaningful

File: config/index.conf.php (hunk 1 above)
Comment (boajer) [reply config/index.conf.php]: Solved

File: config/index.conf.php [hunk 2]
```diff
@@ -0,0 +1,328 @@
...
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ *
+ * Copyright (c) 2020 (original work) Open Assessment Technologies SA;
+ */
+
+declare(strict_types=1);
+
+$dynamicTemplates = [
+    [
+        "HTMLArea" => [
```
Comment (shpran) [inline config/index.conf.php:24]: Please, use single quotes

File: src/Action/IndexCreator.php [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+use oat\oatbox\extension\script\ScriptAction;
+use oat\tao\elasticsearch\ElasticSearch;
+use oat\tao\model\search\Search;
+
+/**
+ * @deprecated  tobe moved to core during upcoming huge refactroing
+ */
+class IndexCreator extends ScriptAction
+{
+
```
Comment (shpran) [inline src/Action/IndexCreator.php:36]:
```suggestion
```

File: src/Action/IndexCreator.php [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+    {
+        return
+            [
+                self::INDEX_FILES => [
+                    'prefix' => 'f',
+                    'longPrefix' => self::INDEX_FILES,
+                    'required' => true,
+                    'description' => 'Absolute path to indices declaration.',
+                ],
+            ];
```
Comment (shpran) [inline src/Action/IndexCreator.php:49]:
```suggestion
        return [
            self::INDEX_FILES => [
                'prefix' => 'f',
                'longPrefix' => self::INDEX_FILES,
                'required' => true,
                'description' => 'Absolute path to indices declaration.',
            ],
        ];
```

File: src/Action/IndexCreator.php [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ *
+ * Copyright (c) 2020 (original work) Open Assessment Technologies SA;
+ */
+
+declare(strict_types=1);
+
+namespace oat\tao\elasticsearch\Action;
+
+use common_report_Report as Report;
```
Comment (shpran) [inline src/Action/IndexCreator.php:25]: As I know, `common_report_Report` is deprecated.
You can use `oat\oatbox\reporting\Report` instead.

File: src/Action/IndexCreator.php [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+                    Report::TYPE_ERROR,
+                    sprintf(
+                        'Error while indices creation: %s',
+                        $exception->getMessage()
+                    )
+                );
+            }
+            return new Report(Report::TYPE_SUCCESS, 'Elastic indexes created successfully');
+        }
+        return new Report(Report::TYPE_ERROR, 'No proper service found');
```
Comment (shpran) [inline src/Action/IndexCreator.php:75]: You can use static methods:
* `Report::createError('message');`
* `Report::createSuccess('message');`

File: src/Action/InitElasticSearch.php [hunk 7]
```diff
@@ -106,38 +115,40 @@ public function __invoke($params)
...
+            $search->createIndexes();
+            $this->getServiceManager()->register(Search::SERVICE_ID, $search);
+
+            $report->add(new Report(Report::TYPE_SUCCESS, __('Switched search service implementation to ElasticSearch')));
+
+            $this->getServiceManager()->register(IndexUpdaterInterface::SERVICE_ID, new IndexUpdater($config['hosts']));
+        } catch (BadRequest400Exception $e) {
+            $report->add(new Report(Report::TYPE_ERROR, 'Unable to create index: ' . $e->getMessage()));
+        } catch (\Exception $e) {
+            $report->add(new Report(Report::TYPE_ERROR, 'ElasticSearch server could not be found'));
```
Comment (shpran) [inline src/Action/InitElasticSearch.php:139]: Same here - you can use static methods to create `Report` with the necessary type

File: src/ElasticSearch.php [hunk 8]
```diff
@@ -33,33 +37,69 @@
...
-    /**
-     * @return \Elasticsearch\Client
-     */
-    protected function getClient()
+    /** @var QueryBuilder */
+    private $queryBuilder;
+
+    public function search(Query $query): SearchResult
+    {
+        $query = [
```
Comment (shpran) [inline src/ElasticSearch.php:50]: It's a bit confusing, because we have an argument with the same name 🙂

File: src/ElasticSearchIndexer.php [hunk 9]
```diff
@@ -29,115 +33,114 @@
...
-    {
-        return $this->client;
+        if (!isset($documentBody['type'])) {
+            throw new RuntimeException('type property is undefined on the document');
+        }
+
+        $documentType = is_string($documentBody['type']) ? [$documentBody['type']] : $documentBody['type'];
+
+        foreach (self::AVAILABLE_INDEXES as $ontology => $indexName) {
+            if (in_array($ontology, $documentType)) {
```
Comment (shpran) [inline src/ElasticSearchIndexer.php:66]: If its possible, use strict comparison

File: src/QueryBuilder.php [hunk 10]
```diff
@@ -0,0 +1,238 @@
...
+        'SearchTextBox',
+    ];
+
+    public static function create(): self
+    {
+        return new self();
+    }
+
+    public function getSearchParams(string $queryString, string $type, int $start, int $count, string $order, string $dir): array
+    {
```
Comment (shpran) [inline src/QueryBuilder.php:80]:
```suggestion
    public function getSearchParams(
        string $queryString,
        string $type,
        int $start,
        int $count,
        string $order,
        string $dir
    ): array {
```

File: src/QueryBuilder.php [hunk 11]
```diff
@@ -0,0 +1,238 @@
...
+                $conditions[] = $this->buildCustomConditions($queryBlock);
+            }
+        }
+
+        return $conditions;
+    }
+
+    private function isStandardField(string $field): bool
+    {
+        return in_array(strtolower($field), self::STANDARD_FIELDS);
```
Comment (shpran) [inline src/QueryBuilder.php:175]:
```suggestion
        return in_array(strtolower($field), self::STANDARD_FIELDS, true);
```

File: src/QueryBuilder.php [hunk 12]
```diff
@@ -0,0 +1,238 @@
...
+        return sprintf(
+            '(%s:("%s"))',
+            self::READ_ACCESS_FIELD,
+            implode('" OR "', $conditions)
+        );
+    }
+
+    private function includeAccessData(string $index): bool
+    {
+        if (!in_array($index, IndexerInterface::INDEXES_WITH_ACCESS_CONTROL)) {
```
Comment (shpran) [inline src/QueryBuilder.php:217]:
```suggestion
        if (!in_array($index, IndexerInterface::INDEXES_WITH_ACCESS_CONTROL, true)) {
```

File: src/QueryBuilder.php [hunk 13]
```diff
@@ -0,0 +1,238 @@
...
+        return $permissionProvider instanceof ReverseRightLookupInterface;
+    }
+
+    private function parseBlock(string $block): QueryBlock
+    {
+        if (\common_Utils::isUri($block)) {
+            return new QueryBlock(null, $block);
+        }
+
+        preg_match('/((?P<field>[^:]*):)?(?P<term>.*)/', $block,$matches);
```
Comment (shpran) [inline src/QueryBuilder.php:232]:
```suggestion
        preg_match('/((?P<field>[^:]*):)?(?P<term>.*)/', $block, $matches);
```

**Review by shpran — APPROVED** [review]
LGTM 👍🏻 
Left several CS comments

Good job!

- [x] New code is covered by tests (if applicable)
- [ ] Tests are running successfully (old and new ones) on my local machine (if applicable)
- [ ] New code is respecting code style rules
- [ ] New code is respecting best practices
- [x] New code is not subject to concurrency issues (if applicable)
- [ ] Feature is working correctly on my local machine (if applicable)
- [x] Acceptance criteria are respected
- [x] Pull request title and description are meaningful

**Pull request merged**